### PR TITLE
Prefer method name rather than return type for iterables

### PR DESCRIPTION
**Pull request by liarco**
Fixes rectorphp/rector#6753

Hi, after following the suggestions from @samsonasik (https://github.com/rectorphp/rector/issues/6753#issuecomment-946438677) I dived into the implementation to understand it a bit further and I ended up with a different fix that I hope you will find useful.
I wanted to improve the solution because the suggested one skips the variable entirely, while it would be better to rename it with a proper name.

If you prefer I can stick to the suggested solution, just let me know!

#### Before:
- when generating the expected name for a foreach variable, a first option based on the returned type is generated
- this option always takes precedence over the one extracted from the method name

This works with the current test cases, but breaks with collective names or other words which are typically used for iterables (e.g. "Collection", "Traversable", etc.)

#### After
- the name based on the returned type is generated, but only returned if `$returnedType` is not ~~"iterable"~~ an array
- the implementation follows as before for the method name, except that when no name can be extracted form the method, then we fallback to the one base on the returned type (if available)

The new implementation works with all previous test cases, but supports my new one where `foreach ($this->getBooks() as $item) {` is correctly changed to `foreach ($this->getBooks() as $book) {` instead of `foreach ($this->getBooks() as $myCollection) {`.

### Stuff to be fixed
~~I only have one issue: PHPStan complains about cognitive complexity, I tried my best and I got it down to 10 (from 12), but I can't find a way to take it down to 9... any advice?~~

Thank you for your time.

### CODE REVIEW

File: rules/Naming/Naming/ExpectedNameResolver.php [hunk 1]
```diff
@@ -168,17 +168,17 @@ public function resolveForForeach(MethodCall | StaticCall | FuncCall $expr): ?st
 
         $expectedNameFromType = $this->propertyNaming->getExpectedNameFromType($returnedType);
 
-        if ($expectedNameFromType !== null) {
+        if ($returnedType->isIterable()->no() && $expectedNameFromType !== null) {
```
Comment (samsonasik) [inline rules/Naming/Naming/ExpectedNameResolver.php]: You can reduce complexity by move to separate private method, eg:

```php
private function isNotIterableAndExpectedNameFromTypeNotNull(\PHPStan\Type\Type $returnedType, ?\Rector\Naming\ValueObject\ExpectedName $expectedNameFromType)
{
       return $returnedType->isIterable()->no() && $expectedNameFromType !== null;
}
```

so you can call:

```php
if ($this-> isNotIterableAndExpectedNameFromTypeNotNull($returnedType, $expectedNameFromType) {
```

Comment (liarco) [reply rules/Naming/Naming/ExpectedNameResolver.php]: That's right, thank you!
I was afraid to pollute the class with random methods. 😅 I'm gonna add this!

**Comment by samsonasik** [comment]
I tried your change locally and try run against `rules` directory and got output:

```diff
bin/rector process rules --dry-run
 868/868 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
5 files with changes
====================

1) rules/DependencyInjection/Rector/Variable/ReplaceVariableByPropertyFetchRector.php:90

    ---------- begin diff ----------
@@ @@
             return null;
         }

-        foreach ($this->variablesToPropertyFetchCollection->getVariableNamesAndTypes() as $name => $type) {
+        foreach ($this->variablesToPropertyFetchCollection->getVariableNamesAndTypes() as $name => $variableNamesAndType) {
             if (! $this->isName($node, $name)) {
                 continue;
             }

-            /** @var ObjectType $type */
-            if (! $this->isObjectType($node, $type)) {
+            /** @var ObjectType $variableNamesAndType */
+            if (! $this->isObjectType($node, $variableNamesAndType)) {
                 continue;
             }
    ----------- end diff -----------

Applied rules:
 * RenameForeachValueVariableToMatchMethodCallReturnTypeRector


2) rules/Naming/Naming/ExpectedNameResolver.php:183

    ---------- begin diff ----------
@@ @@
         return $expectedNameFromMethodName->getSingularized();
     }

-    private function isNotIterableAndExpectedNameFromTypeNotNull(
-        Type $returnedType,
-        ?ExpectedName $expectedNameFromType
-    ): bool {
-        return $returnedType->isIterable()
-            ->no() && $expectedNameFromType !== null;
+    private function isNotIterableAndExpectedNameFromTypeNotNull(Type $returnedType, ?ExpectedName $expectedName): bool
+    {
+        if (!$returnedType->isIterable()
+            ->no()) {
+            return false;
+        }
+        return $expectedName !== null;
     }

     private function isDynamicNameCall(MethodCall | StaticCall | FuncCall $expr): bool
    ----------- end diff -----------

Applied rules:
 * RenameParamToMatchTypeRector


3) rules/Php70/Rector/FuncCall/NonVariableToVariableOnFunctionCallRector.php:131

    ---------- begin diff ----------
@@ @@
             return [];
         }

-        foreach ($functionLikeReflection->getVariants() as $parametersAcceptor) {
+        foreach ($functionLikeReflection->getVariants() as $variant) {
             /** @var ParameterReflection $parameterReflection */
-            foreach ($parametersAcceptor->getParameters() as $key => $parameterReflection) {
+            foreach ($variant->getParameters() as $key => $parameterReflection) {
                 // omitted optional parameter
                 if (! $this->argsAnalyzer->isArgInstanceInArgsPosition($call->args, $key)) {
                     continue;
    ----------- end diff -----------

Applied rules:
 * RenameForeachValueVariableToMatchMethodCallReturnTypeRector


4) rules/Php71/Rector/FuncCall/RemoveExtraParametersRector.php:117

    ---------- begin diff ----------
@@ @@
         MethodReflection | FunctionReflection $functionLikeReflection
     ): int {
         $parameterCounts = [0];
-        foreach ($functionLikeReflection->getVariants() as $parametersAcceptor) {
-            $parameterCounts[] = count($parametersAcceptor->getParameters());
+        foreach ($functionLikeReflection->getVariants() as $variant) {
+            $parameterCounts[] = is_countable($variant->getParameters()) ? count($variant->getParameters()) : 0;
         }

         return (int) max($parameterCounts);
    ----------- end diff -----------

Applied rules:
 * FunctionArgumentDefaultValueReplacerRector (https://php.watch/versions/8.1/version_compare-operator-restrictions)
 * RecastingRemovalRector
 * MysqlFuncCallToMysqliRector (https://www.phpclasses.org/blog/package/9199/post/3-Smoothly-Migrate-your-PHP-Code-using-the-Old-MySQL-extension-to-MySQLi.html)
 * RenameForeachValueVariableToMatchMethodCallReturnTypeRector
 * RemoveReferenceFromCallRector
 * RandomFunctionRector
 * CountOnNullRector
 * RemoveFuncCallArgRector


5) rules/Transform/NodeTypeAnalyzer/TypeProvidingExprFromClassResolver.php:75

    ---------- begin diff ----------
@@ @@
         ClassReflection $classReflection,
         ObjectType $objectType
     ): ?MethodCall {
-        foreach ($classReflection->getNativeMethods() as $methodReflection) {
-            $functionVariant = ParametersAcceptorSelector::selectSingle($methodReflection->getVariants());
+        foreach ($classReflection->getNativeMethods() as $nativeMethod) {
+            $functionVariant = ParametersAcceptorSelector::selectSingle($nativeMethod->getVariants());
             $returnType = $functionVariant->getReturnType();

             if (! $this->isMatchingType($returnType, $objectType)) {
@@ @@
             }

             $thisVariable = new Variable('this');
-            return new MethodCall($thisVariable, $methodReflection->getName());
+            return new MethodCall($thisVariable, $nativeMethod->getName());
         }

         return null;
    ----------- end diff -----------

Applied rules:
 * RenameForeachValueVariableToMatchMethodCallReturnTypeRector


                                                                                                                        
 [OK] 5 files would have changed (dry-run) by Rector 
```

for `key` -> `valueVar`, it seems it change that already ok, for example:

```diff
-        foreach ($this->variablesToPropertyFetchCollection->getVariableNamesAndTypes() as $name => $type) {
+        foreach ($this->variablesToPropertyFetchCollection->getVariableNamesAndTypes() as $name => $variableNamesAndType) {
```

I think for has `key` and `valueVar` need to be kept as the previous, what do you think?

Also, this seems already previously correct:

```diff
-        foreach ($classReflection->getNativeMethods() as $methodReflection) {
-            $functionVariant = ParametersAcceptorSelector::selectSingle($methodReflection->getVariants());
+        foreach ($classReflection->getNativeMethods() as $nativeMethod) {
```
You can try other directory rectify command with list of the dirs at 

https://github.com/rectorphp/rector-src/blob/3ba32894dfebb1e746c8d1c1520b2d792145308a/.github/workflows/rector.yaml#L28-L30

**Comment by TomasVotruba** [comment]
@samsonasik Feel free to merge, when you feel this is ready :+1:

**Comment by liarco** [comment]
Thank you @samsonasik, I just run `composer complete-check`, I didn't know about this extra step.

By the way I think that I missed [this line](https://github.com/rectorphp/rector-src/blob/main/rules/Naming/Naming/ExpectedNameResolver.php#L163) where the actual returned type is replaced by the inner type when it's an array... I need to check this in order to consider all the possible options. I'm not sure but maybe I should add more tests to cover more cases.

**Comment by liarco** [comment]
@samsonasik I updated my code and now it seems to work properly. (I created a separate variable for `$returnedType` and `$innerReturnedType` in order to avoid confusion in the future)

I run rector on all the suggested paths and it doesn't change anything except the following:
```diff
bin/rector process src tests rules-tests packages packages-tests --dry-run
 1713/1713 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
1 file with changes
===================

1) tests/PhpUnit/MultipleFilesChangedTrait/Rector/Class_/CreateJsonWithNamesForClassRector.php:30

    ---------- begin diff ----------
@@ @@

     /**
      * @param Class_ $node
+     * @return null
      */
     public function refactor(Node $node)
     {
    ----------- end diff -----------

Applied rules:
 * AddArrayReturnDocTypeRector


                                                                                                                        
 [OK] 1 file would have changed (dry-run) by Rector                                                                     
                                                                                                                        


```

But I don't think that this is related to my changes.

--------

I'm also working on trying to extend this rule in order to do the same on both arrays and iterable objects that implement generics (e.g. `getValues(): Book[]` and `getValues(): Collection // with @return Collection<Book>` should both use `$book` as variable). But maybe this should be left for another PR, wdyt?

**Comment by samsonasik** [comment]
@liarco thank you, yes, that can be in separate PR.

**Comment by samsonasik** [comment]
Thank you @liarco

**Pull request merged**